pkg/controller/argocdmetrics/argocd_metrics_controller.go

```diff
@@ -0,0 +1,409 @@
+		err = r.client.Update(context.TODO(), &namespace)
+		if err != nil {
+			reqLogger.Error(err, "Error updating namespace",
+				"Namespace", namespace.Name)
+			return reconcile.Result{}, err
+		}
+	} else {
+		reqLogger.Info("Namespace already has cluster-monitoring label",
+			"Namespace", namespace.Name)
+	}
+
+	// Create role to grant read permission to the openshift metrics stack
+	err = r.createReadRoleIfAbsent(request.Namespace, argocd, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create role binding to grant read permission to the openshift metrics stack
+	err = r.createReadRoleBindingIfAbsent(request.Namespace, argocd, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create ServiceMonitor for ArgoCD application metrics
+	serviceMonitorLabel := fmt.Sprintf("%s-metrics", request.Name)
+	serviceMonitorName := request.Name
+	err = r.createServiceMonitorIfAbsent(request.Namespace, argocd, serviceMonitorName, serviceMonitorLabel, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create ServiceMonitor for ArgoCD API server metrics
+	serviceMonitorLabel = fmt.Sprintf("%s-server-metrics", request.Name)
+	serviceMonitorName = fmt.Sprintf("%s-server", request.Name)
+	err = r.createServiceMonitorIfAbsent(request.Namespace, argocd, serviceMonitorName, serviceMonitorLabel, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create ServiceMonitor for ArgoCD repo server metrics
+	serviceMonitorLabel = fmt.Sprintf("%s-repo-server", request.Name)
+	serviceMonitorName = fmt.Sprintf("%s-repo-server", request.Name)
+	err = r.createServiceMonitorIfAbsent(request.Namespace, argocd, serviceMonitorName, serviceMonitorLabel, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create alert rule
+	err = r.createPrometheusRuleIfAbsent(request.Namespace, argocd, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ArgoCDMetricsReconciler) createReadRoleIfAbsent(namespace string, argocd *argoapp.ArgoCD, reqLogger logr.Logger) error {
+	readRole := newReadRole(namespace)
+	existingReadRole := &rbacv1.Role{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: readRole.Name, Namespace: readRole.Namespace}, existingReadRole)
+	if err == nil {
+		reqLogger.Info("Read role already exists",
+			"Namespace", readRole.Namespace, "Name", readRole.Name)
+		return nil
+	}
+	if errors.IsNotFound(err) {
+		reqLogger.Info("Creating new read role",
+			"Namespace", readRole.Namespace, "Name", readRole.Name)
+
+		// Set the ArgoCD instance as the owner and controller
+		if err := controllerutil.SetControllerReference(argocd, readRole, r.scheme); err != nil {
+			reqLogger.Error(err, "Error setting read role owner ref",
+				"Namespace", readRole.Namespace, "Name", readRole.Name, "ArgoCD Name", argocd.Name)
+			return err
+		}
+
+		err = r.client.Create(context.TODO(), readRole)
```

Comment:
Just curious if this is normally the correct order to set controlllerReference before create the reousrce.